### PR TITLE
Add Heap Record-Replay, -memory record

### PR DESCRIPTION
**Pull request by evan-charmworks**
Generates an MPI program that repeats the exact sequence of heap operations observed on each PE. Can be used for debugging and performance analysis of memory allocators as well as heap behavior of user codes.

### CODE REVIEW

File: src/conv-core/memory-record.C [hunk 1]
```diff
@@ -0,0 +1,329 @@
...
+  if (CmiMyRank() == 0)
+    initialized = 1;
+
+  CmiNodeAllBarrier();
+
+  if (mype == 0)
+  {
+    const int numpes = CmiNumPes();
+
+    FILE * mainsrc = fopen("heap-replay.cpp", "wb");
```
Comment (matthiasdiener) [inline src/conv-core/memory-record.C]: should probably be 'w' instead of 'wb' (same for the other fopen())

**Comment by matthiasdiener** [comment]
This is an interesting idea. Can you add more information about possible use cases for this?

File: src/conv-core/memory-record.C (hunk 1 above)
Comment (evan-charmworks) [reply src/conv-core/memory-record.C]: Done

**Comment by evan-charmworks** [comment]
> This is an interesting idea. Can you add more information about possible use cases for this?

Added. I wrote this to help diagnose what Laghos was doing that caused the Isomempool to perform so poorly. By inspecting the output files I could see that it was allocating and then freeing the same two sequences of buffers tens of thousands of times.

**Pull request merged**